autoscaler/collectormetrics/autoscaler.go

```diff
@@ -0,0 +1,147 @@
+package collectormetrics
+
+import (
+	"time"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/odigos-io/odigos/k8sutils/pkg/env"
+
+	"github.com/odigos-io/odigos/autoscaler/controllers/datacollection"
+	"github.com/odigos-io/odigos/autoscaler/controllers/gateway"
+
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	odigosv1 "github.com/odigos-io/odigos/api/odigos/v1alpha1"
+)
+
+type AutoscalerOptions struct {
+	interval        time.Duration
+	collectorsGroup odigosv1.CollectorsGroupRole
+	minReplicas     int
+	maxReplicas     int
+	algorithm       AutoscalerAlgorithm
+}
+
+type AutoscalerOption func(*AutoscalerOptions)
+
+func WithInterval(interval time.Duration) AutoscalerOption {
+	return func(o *AutoscalerOptions) {
+		o.interval = interval
+	}
+}
+
+func WithScaleRange(minReplicas, maxReplicas int) AutoscalerOption {
+	return func(o *AutoscalerOptions) {
+		o.minReplicas = minReplicas
+		o.maxReplicas = maxReplicas
+	}
+}
+
+func WithCollectorsGroup(collectorsGroup odigosv1.CollectorsGroupRole) AutoscalerOption {
+	return func(o *AutoscalerOptions) {
+		o.collectorsGroup = collectorsGroup
+	}
+}
+
+func WithAlgorithm(algorithm AutoscalerAlgorithm) AutoscalerOption {
+	return func(o *AutoscalerOptions) {
+		o.algorithm = algorithm
+	}
+}
+
+type Autoscaler struct {
+	kubeClient    client.Client
+	options       AutoscalerOptions
+	ticker        *time.Ticker
+	notifications chan Notification
+	podIPs        map[string]string
```

Comment:
Why this map needs to have the pod names as values? Can we just use a `map[string]struct{}` which will only save the IPs?

Reply:
I am using the pod name just in case of an error.
This allows us to print an informative error message that contains the pod name and not just the IP which will be harder to debug.